lib/cn/cn_tree.c

```diff
@@ -1965,20 +1961,263 @@ static merr_t
 cn_subspill_apply(struct subspill *ss)
 {
     struct cn_compaction_work *w = ss->ss_work;
-    merr_t err;
+    merr_t err = 0;
+    bool delete_node = false;
+    struct cn_tree_node *outnode;
+
+    if (!ss->ss_added)
+        return 0;
+
+    if (ss->ss_is_zspill) {
+        struct kvset_list_entry *le;
+
+        err = cn_move(w, w->cw_node, ss->ss_zspill.zsp_src_list,
+                      w->cw_kvset_cnt, delete_node, ss->ss_node);
+
+        w->cw_output_nodev = &outnode;
+
+        if (err) {
+            uint i = 0;
+
+            list_for_each_entry(le, &w->cw_node->tn_kvset_list, le_link) {
+                assert(kvset_get_workid(le->le_kvset) != 0);
+                kvset_set_workid(le->le_kvset, 0);
+
+                if (++i >= w->cw_kvset_cnt)
+                    break;
+            }
+        }
+
+    } else {
+        err = cn_subspill_commit(ss);
+    }
 
-    err = cn_subspill_commit(ss);
     if (err)
         return err;
 
-    if (ss->ss_added) {
-        w->cw_output_nodev[0] = ss->ss_node;
-        w->cw_checkpoint(w);
-    }
+    w->cw_output_nodev[0] = ss->ss_node;
+    w->cw_checkpoint(w);
 
     return 0;
 }
 
+/* Tree must be locked when this is called */
+struct cn_tree_node *
+cn_kvset_can_zspill(struct kvset *ks, struct route_map *map)
+{
+    unsigned char maxkbuf[HSE_KVS_KEY_LEN_MAX];
+    const void *maxkey, *maxptkey, *minkey;
+    uint16_t maxklen, maxptklen, minklen;
+
+    struct route_node *minnode;
+
+    kvset_minkey(ks, &minkey, &minklen);
+    kvset_maxkey(ks, &maxkey, &maxklen);
+    kvset_max_ptkey(ks, &maxptkey, &maxptklen);
+
+    if (maxptkey && keycmp_prefix(maxptkey, maxptklen, maxkey, maxklen) >= 0) {
+        maxklen = maxptklen;
+
+        memcpy(maxkbuf, maxptkey, maxptklen);
+        memset(&maxkbuf[maxptklen], 0xff, sizeof(maxkbuf) - maxptklen);
+        maxkey = maxkbuf;
+    }
+
+    minnode = route_map_lookup(map, minkey, minklen);
+    assert(minnode);
+
+    if (route_node_keycmp(minnode, maxkey, maxklen) >= 0)
+        return route_node_tnode(minnode);
+
+    return NULL;
+}
+
+/* TODO: This function shares a fair bit of code with cn_comp_spill() and should be refactored.
+ */
+static merr_t
+cn_zspill(struct cn_compaction_work *w, bool *zspilled)
+{
+    struct kvset_list_entry *first, *le = w->cw_mark;
+    struct cn_tree *tree = w->cw_tree;
+    struct cn_tree_node *n, *znode;;
+    struct subspill *ss_arr;
+    struct route_node *rtn = NULL;
+    atomic_uint *spillingp = NULL;
+    atomic_uint *zspillingp = NULL;
+    void *lock;
+    uint i;
+    merr_t err = 0;
```

Comment:
We use "tn" for a locate tree node ptr everywhere in cn, except a few places where "node" has popped up (and should be changed to tn).  So, change this "n" to "tn" please.  Not to mention that "n" is a classic name for a counting integer..